Add 'profile' option to 'cache warm' command

mollyIV opened the pull request:
Resolves https://github.com/tuist/tuist/issues/2190

### Short description 📝

This pull request is adding a `profile` option to `cache warm` command:

```bash
$ tuist cache warm --profile simulator
```

More details in [linked issue's description](https://github.com/tuist/tuist/issues/2190).

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [x] A developer other than the author has verified that the changes work as expected.
- [x] The changes have been tested following the [documented guidelines](https://tuist.io/docs/contribution/testing-strategy/).
- [x] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [x] In case the PR introduces changes that affect users, the documentation has been updated.

mollyIV commented:
FYI: The `Next Website / Build` job is failing on a `main` branch too.

fortmarek commented on CHANGELOG.md:
```diff
@@ -20,6 +20,7 @@ Please, check out guidelines: https://keepachangelog.com/en/1.0.0/
 - Add support for embedded scripts in a TargetAction. [#2192](https://github.com/tuist/tuist/pull/2192) by [@jsorge]
 - Support `.s` source files [#2199](https://github.com/tuist/tuist/pull/2199) by[ @dcvz](https://github.com/dcvz).
 - Support for printing from the manifest files [#2215](https://github.com/tuist/tuist/pull/2215) by @pepibumur](https://github.com/pepibumur).
+- Add 'configuration' option to 'cache warm' command [#2190](https://github.com/tuist/tuist/issues/2190) by [@mollyIV](https://github.com/mollyIV).
```
```suggestion
- Add `configuration` option to `cache warm` command [#2190](https://github.com/tuist/tuist/issues/2190) by [@mollyIV](https://github.com/mollyIV).
```

fortmarek commented on features/step_definitions/cache.rb:
```diff
@@ -8,6 +8,10 @@
 end
 
+Then(/^tuist warms the cache with configuration ([a-zA-Z]+)$/) do |configuration|
+  system("swift", "run", "tuist", "cache", "warm", "--path", @dir, "--configuration", configuration)
```
thanks for doing this 👏

fortmarek commented on fixtures/ios_app_with_custom_configuration/App/Sources/AppDelegate.swift:
```diff
@@ -0,0 +1,28 @@
+        
+        FrameworkA.hello()
+        
+        #if DEBUG_MACRO
```
why not just use `#if DEBUG` rather than defining a custom macro?

mollyIV replied:
The reason behind choosing a custom debug compiler macro `DEBUG_MACRO` over the pre-defined `DEBUG` macro was that previously, `Debug` configuration was used by default, so a `DEBUG` macro was available. Having unique, custom compiler macro in custom configuration proofs that the changes work correctly, no related to default settings. Does that make sense? 😊

fortmarek replied:
yes! thanks

fortmarek reviewed (approved):
Just a few nitpicks - I have tried the changes on our project and it seems to have worked great 🤗

pepibumur commented:
👋 I'm chiming in here. I agree with the idea of giving developers the flexibility to choose how they'd like to build their cacheable artifacts:

- Framework or .xcframework?
- Build configuration
- Device, Simulator, Both?

However, since those arguments need to be accounted for when hashing the targets, developers might get unexpected results if they don't use the same arguments with focus as the ones that were used for warming the cache. For example. If on CI we warm with the "Debug" configuration, but locally with focus on the "Release" configuration, developers won't get pre-compiled targets and they won't understand why. This leads to developers thinking it's a bug on Tuist.

Like Xcode does with configuration, I think we can introduce the notion of cache configurations/flavors/profiles. Here's some pseudo-code of how the API might look like:

```swift
let config = Config(cache: .cache(profiles: [.profile(name: "Simulator", configuration: "Debug", product: .xcframework)])
```

One of the flavors would be the default. If developers don't specify the flavor, we take the default one. Thanks to that, teams can scope their caching profiles and minimize the list of arguments developers would need to pass:

```bash
tuist warm cache --flavor development
tuist focus Target --flavor development
```


cc @andreacipriani / @RomainBoulay

mollyIV commented:
Hey @pepibumur 👋 

Thanks for sharing this idea. I really like it and I do believe introducing the profiles / flavors will allow to avoid confusions when using `tuist`'s CLI. 

The snippet you provided looks decent. It can be easily extendable, i.e. adding more options like a device or os. In terms of naming, I'd vote for `flavor`, because `profile` is quite reserved in iOS world and may be mistaken with provisioning profiles.

```bash
let config = Config(cache: .cache(flavors: [.flavor(name: "Simulator", configuration: "Debug", product: .xcframework)])
```

Follow-up questions:

1. What about `tuist generate` command - shouldn't that be in parity with `tuist cache` command too?
2. Should a `product` be a part of cache `flavor`? Please note that currently there's `--xcframeworks` flag available for `tuist cache warm` command. 

🙇 🙇

mollyIV commented on Sources/ProjectDescription/Config.swift:
```diff
@@ -34,23 +34,29 @@ public struct Config: Codable, Equatable {
     public init(
         compatibleXcodeVersions: CompatibleXcodeVersions = .all,
         cloud: Cloud? = nil,
+        cache: Cache? = nil,
```
I was wondering if we should provide the default value for the config at the project's manifest level 🤔 For now, the default value of a config is provided at the `TuistCore.Cache` level. 

Thoughts?

pepibumur replied:
This is the behavior that I'd expect as a user:

- Passed flavor exists: It uses it  and tuist prints the flavor
- Passed flavor doesn't exist: It fails the execution.
- No flavor is passed and there are flavors defined: It picks the one from the list, Tuist's output tells me that it's taking the default one.
- No flavor is passed and the list of flavors is empty: It defaults to Tuist's defaults and tells me what the defaults are.

mollyIV replied:
Awesome, I added unit tests to verify described scenarios 💪

mollyIV commented on Sources/TuistCache/Utilities/CacheFrameworkBuilder.swift:
```diff
@@ -65,26 +65,31 @@ public final class CacheFrameworkBuilder: CacheArtifactBuilding {
 
     fileprivate func build(_ projectTarget: XcodeBuildTarget,
                            target: Target,
+                           configuration: String?,
```
Eventually, `configuration` argument will be non-optional, because a `configuration` will always be provided via default config.

mollyIV replied:
Changed.

mollyIV commented on Sources/TuistKit/Services/Cache/CacheWarmService.swift:
```diff
@@ -41,7 +43,50 @@ final class CacheWarmService {
     }
 }
+
+enum CacheWarmServiceError: FatalError, Equatable {
```
I am opened for the suggestions related to error cases naming and descriptions 😊

pepibumur replied:
One suggestion that I'd make is moving this to the top. It's been a non-written convention.

mollyIV commented on Sources/TuistLoader/Models+ManifestMappers/Config+ManifestMapper.swift:
```diff
@@ -19,8 +19,14 @@ extension TuistCore.Config {
+        }
         return TuistCore.Config(compatibleXcodeVersions: compatibleXcodeVersions,
                                 cloud: cloud,
+                                cache: cache ?? .default,
```
I am not sure, if providing a default cache in a mapper is a good place of doing so, thoughts? 🤔

pepibumur replied:
I suggested this in another comment but yeah, I think it should be done in the `CacheWarmService`.

mollyIV replied:
Will do 👍

mollyIV commented:
I started working on adding `flavor` option to `cache warm` command (according to [Pedro's suggestion](https://github.com/tuist/tuist/pull/2221#issuecomment-754055151)). I pushed the part 1 and  added some follow-up questions in the comments. I'd appreciate the clarifications 😊  In the meantime, I am going to work on the next steps. Thank you!

mollyIV commented on features/cache-frameworks.feature:
```diff
@@ -18,4 +18,12 @@ Feature: Focuses projects with pre-compiled cached xcframeworks
     When tuist focuses the target App
+    Then I should be able to build for iOS the scheme App
+
+  Scenario: The project is an iOS application with custom configuration and cache flavor (ios_app_with_custom_configuration)
```
What's surprising the acceptance test is failing due to 

```bash
$ ./bin/fourier test tuist acceptance features/cache-frameworks.feature:23
... 
Then tuist warms the cache with flavor simulator                                                                         # features/step_definitions/cache.rb:11
      swift run tuist cache warm --path /var/folders/9p/kwbtgrj53cl08_1gwbtm5kvc0000gn/T/d20210108-62798-1ottqko --flavor simulator
      The flavor 'simulator' is missing in your project's configuration. Available cache flavors: development.
```

It seems that a cache config is not loaded from a `Config.swift` manifest file and a default config is used instead. Does anyone have an idea why?

pepibumur replied:
Does the test run locally?

mollyIV replied:
Apparently, the test is passing after applying the changes 👍

mollyIV replied:
Now, when I execute
```bash
./bin/fourier test tuist acceptance features/cache-frameworks.feature:23
```

I got:

```bash
Then tuist warms the cache with simulator profile                                                                         # features/step_definitions/cache.rb:11
      swift run tuist cache warm --path /var/folders/9p/kwbtgrj53cl08_1gwbtm5kvc0000gn/T/d20210127-62892-ngaxvs --profile simulator
      Couldn't find framework 'FrameworkA.framework' in the derived data directory: /Users/djankowski/Library/Developer/Xcode/DerivedData
       (Minitest::Assertion)
      ./features/support/system.rb:9:in `system'
      ./features/step_definitions/cache.rb:12:in `/^tuist warms the cache with ([a-zA-Z]+) profile/'
      features/cache-frameworks.feature:27:in `tuist warms the cache with simulator profile'
```

When I execute the test via terminal myself, it seems that the cache command does not pick up `Config.swift` config: 

```bash
$ swift run --package-path ~/Projects/tuist tuist cache warm --path ~/Projects/tuist/fixtures/ios_app_with_custom_configuration --profile simulator
Selected cache profile 'development' from Tuist's defaults // <--- although the simulator profile is specified in Config
Generating workspace Workspace.xcworkspace
Generating project App
Generating project FrameworkA
Hashing cacheable targets
The target FrameworkA with hash 97c2c7c2b25f87c168306d21b1e83fa7 and type framework is already in the cache. Skipping...
Building cacheable targets
All cacheable targets have been cached successfully as frameworks
```

Any ideas @pepibumur ? 🙏

pepibumur commented on Sources/ProjectDescription/Cache.swift:
```diff
@@ -0,0 +1,20 @@
+import Foundation
+
+// TODO: Add documentation
```
😬 We should address those before merging the PR.

mollyIV replied:
Absolutely, I am going to tackle all TODOs I added and fill in the documentation.

pepibumur commented on Sources/ProjectDescription/Cache.swift:
```diff
@@ -0,0 +1,20 @@
+// TODO: Add documentation
+// TODO: Add unit tests
+public struct Cache: Codable, Equatable {
+    public struct Flavor: Codable, Equatable {
```
Nitpick: I'm more inclined towards calling this one **profile**. What do you think?

mollyIV replied:
Sounds good 👍 Let me rename it then.

pepibumur commented on Sources/ProjectDescription/Cache.swift:
```diff
@@ -0,0 +1,20 @@
+
+    public let flavors: [Flavor]
+
+    public static func cache(flavors: [Flavor]) -> Cache {
```
I'd mention in the documentation that the first in the list will be taken as the default if no flavor is provided.

pepibumur commented on Sources/TuistCore/Cache/Cache.swift:
```diff
@@ -0,0 +1,21 @@
+        self.flavors = flavors
+    }
+
+    public static let `default` = Cache(flavors: [Flavor(name: "development", configuration: "Debug")])
```
There's a caveat in this approach. It's possible that the `Debug` configuration doesn't exist. This is very unlikely, but I was wondering of deciding the default should be something that we do later in the process.

pepibumur commented on Sources/TuistKit/Commands/Cache/CacheWarmCommand.swift:
```diff
@@ -17,13 +17,19 @@ struct CacheWarmCommand: ParsableCommand {
+        name: .shortAndLong,
+        help: "The name of the flavor to be used when warming up the cache."
+    )
+    var flavor: String?
```
As I mentioned earlier, I'd call this one `profile`.

mollyIV replied:
Done 👍

pepibumur commented on Sources/TuistKit/Services/Cache/CacheWarmService.swift:
```diff
@@ -28,7 +29,8 @@ final class CacheWarmService {
             cacheController = cacheControllerFactory.makeForSimulatorFramework(contentHasher: contentHasher)
         }
-        try cacheController.cache(path: path)
+        let flavor = try cacheFlavor(named: flavor, from: config)
```
@mollyIV bear in mind that we need to extend `CacheContentHasher`'s hashing logic to include the values from the flavor configuration. Otherwise 2 different flavors, which yield different binaries, will have the same hash.

mollyIV replied:
Any suggestions where we could add it?

pepibumur commented on Sources/TuistKit/Services/Cache/CacheWarmService.swift:
```diff
@@ -41,7 +43,50 @@ final class CacheWarmService {
+        switch flavorName {
+        case .none:
+            guard let defaultFlavor = flavors.first else {
+                throw CacheWarmServiceError.missingDefaultFlavor
```
Remember to document this on the documentation website. This is a convention that we are defining and that is important to document.

mollyIV replied:
Can you please show me an example on the documentation website? 🤔 Not sure how to document it.

pepibumur commented on Sources/TuistKit/Services/Cache/CacheWarmService.swift:
```diff
@@ -41,7 +43,50 @@ final class CacheWarmService {
+    var description: String {
+        switch self {
+        case let .missingFlavor(name, availableFlavors):
+            return "The flavor '\(name)' is missing in your project's configuration. Available cache flavors: \(availableFlavors.joined(separator: ", "))."
```
I believe [Swift](https://developer.apple.com/documentation/foundation/listformatter) has a cool formatter for formatting lists :)

pepibumur commented on website/markdown/docs/building-at-scale/caching.mdx:
```diff
@@ -27,6 +27,12 @@ tuist cache warm
 <ArgumentsTable
   args={[
+    {
+      long: '`--configuration`',
```
We'll have to update this one.

pepibumur reviewed (commented):
👏 Solid work here @mollyIV

mollyIV commented:
Thanks for the approval @pepibumur  🙏 

Please note that there are still some todos and docs to be updated, so please do not merge it yet. I will let you know when it's ready for the final review 😊 Thanks!

mollyIV commented:
Hey @pepibumur 👋 

Thanks again for your feedback, I addressed all the issues, however I am still struggling with one:
* https://github.com/tuist/tuist/pull/2221#discussion_r565545361. Could you please have a look? 🙏 

The issue might be related to custom schemes and configurations definition in project's manifest + caching.

fortmarek commented on Tests/TuistKitTests/Services/Cache/CacheProfileResolverTests.swift:
```diff
@@ -0,0 +1,125 @@
+        super.tearDown()
+    }
+
+    func testItResolvesDefaultProfileFromTuistDefaults() {
```
nitpick: most of words in a name of test cases are delimited by underscore rather than using camel case

fortmarek commented on website/markdown/docs/building-at-scale/caching.mdx:
```diff
@@ -27,6 +27,12 @@ tuist cache warm
+    {
+      long: '`--profile`',
+      short: '`-P`',
+      description: 'The name of the profile to be used when warming up the cache.',
```
I think the profile deserves its own section on this page. If I just read the website's documentation, I have no idea what a profile actually is

mollyIV replied:
I added the section 👍

fortmarek reviewed (commented):
Profiles are a really good solution! Some nitpicks plus, as mentioned, I'd add a section for what profiles are.

pepibumur reviewed (approved):
After addressing @fortmarek's comments and resolving the conflicts I think we are good to merge.

mollyIV commented on Sources/TuistKit/Services/Cache/CacheProfileResolver.swift:
```diff
@@ -0,0 +1,43 @@
+        // the list of profiles in Config file is empty.
+        guard
+            let name = profileName,
+            let cacheConfig = config.cache,
```
I did the change, please have a look 🙏

andreacipriani commented:
I'm late on this one, but I wanted to say amazing job @mollyIV !!!

fortmarek commented on website/markdown/docs/building-at-scale/caching.mdx:
```diff
@@ -45,6 +51,10 @@ tuist cache warm
 
+### Caching profile 
+
+The caching profile allows users to specify how targets will be cached, for example setting a configuration. Instead of passing the configuration through CLI arguments, developers can define a profile and reference it with `tuist cache warm --profile MyProfile`.
```
As I wrote to you on Slack, I'd give also a quick example of how could a `Profile`, eg:

You can define a Profile in `Config.swift`, for example like this:
```swift
let config = Config(cache: .cache(profiles: [.profile(name: "Simulator", configuration: "Debug",)])
```

mollyIV replied:
Sure, I updated the documentation 👍

fortmarek reviewed (approved):
I have a small nitpick to the documentation, but otherwise it's good from my side 👍

mollyIV commented:
Please do not merge it yet, I am working on the last comment: https://github.com/tuist/tuist/pull/2221#discussion_r557563940.

mollyIV commented:
Hey @pepibumur @fortmarek 👋 

I investigated the amount of work in order to take into account a caching profile (or the plain values included in the profile like configuration, etc.) while calculating a hash. It would require adding a `--profile` option to the following commands / subcommands:

* cache print-hashes
* focus
* generate

The following commands wouldn't need to have a profile option, but we would need to use its options instead, like configuration or a device (when a caching profile supports it):

* build
* doc
* test

The amount of changes will be quite significant. I wouldn't like to make this, already big, PR even bigger, so I'd like to propose removing a `--profile` option, but leaving the work under the hood. Then, we will work on adding a `--profile` option properly - taking it into account while caching. 

Thoughts?

RomainBoulay commented:
Hey @mollyIV!

You probably don't need to alter the `generate` command as cached artefacts won't be used by this command, if I'm not wrong?

That said, I agree there's more work and that it might be hard to continue working on a big and open PR. I'm not sure `tuist` already uses a feature flag system, but here that might be a good solution. My recommendation would be to hide your code behind an assertion (that is false by default - maybe you can use an environment variable?) and merge it. That way, your code is not run by our users until we finish the rest of the work you mentioned. Does that make sense @mollyIV? @pepibumur what do you think?

RomainBoulay commented:
Ho and btw... like Andrea said, that's really great work 👏 Thanks a lot for contributing @mollyIV 🙏

mollyIV commented:
Hey @RomainBoulay 👋 

> You probably don't need to alter the generate command as cached artefacts won't be used by this command, if I'm not wrong?

I double checked the codebase and did a quick prototype and you are right. I updated the list of commands that will require to support a `--profile` options. It seems that there is going to be `focus` and `cache print-hashes` only.

> My recommendation would be to hide your code behind an assertion (that is false by default - maybe you can use an environment variable?) and merge it. 

I was thinking about removing the `--profile` option from `cache warm` command and not exposing a `config` param in a `Config.swift` file. The defaults populated into the code under the hood will work just fine 😉 WDYT?

RomainBoulay commented:
It sounds good to me @mollyIV! Thanks!!

mollyIV commented:
👋 I have hidden the cache profiles from being publicly available. When this pull request is merged, I will open the follow-up pull request and work on including the caching profile into cache hashing 🙇

pepibumur commented:
@mollyIV would you mind rebasing from `main`? Once it's green we can merge.

mollyIV commented:
@pepibumur sure, I did take the newest changes from 'main' branch and all the checks are green 💪
